File: feature_integration_tests/test_scenarios/cpp/src/scenarios/persistency/utf8_defaults.cpp
// *******************************************************************************
// Copyright (c) 2026 Contributors to the Eclipse Foundation
//
// See the NOTICE file(s) distributed with this work for additional
// information regarding copyright ownership.
//
// This program and the accompanying materials are made available under the
// terms of the Apache License Version 2.0 which is available at
// https://www.apache.org/licenses/LICENSE-2.0
//
// SPDX-License-Identifier: Apache-2.0
// *******************************************************************************

#include "../../internals/persistency/kvs_instance.h"

#include <scenario.hpp>

#include <iostream>
#include <stdexcept>
#include <string>

namespace {

/// Open a KVS whose defaults file was created with UTF-8 encoded key names.
/// Override only the emoji key and flush. Python verifies:
/// - the emoji key appears in the snapshot with the override value
/// - the ASCII and Greek keys are absent (never explicitly written)
///
/// Combines feat_req__persistency__support_datatype_keys,
/// feat_req__persistency__default_values, and
/// feat_req__persistency__default_value_file.
class Utf8Defaults : public Scenario {
public:
    std::string name() const override {
        return "utf8_defaults";
    }

    void run(const std::string& input) const override {
        KvsParameters params = KvsParameters::from_json_section(input, "kvs_parameters_1");

        auto kvs_opt = KvsInstance::create(params);
        if (!kvs_opt) {
            throw std::runtime_error("Failed to create KVS instance");
        }
        auto kvs = *kvs_opt;

        // Override only the emoji key; ascii and greek keys remain as defaults.
        if (!kvs->set_value(u8"utf8_emoji 🔑", 777.0)) {
            throw std::runtime_error("Failed to set value");
        }

        if (!kvs->flush()) {
            throw std::runtime_error("Failed to flush KVS");
        }

        // Log default values for ascii and greek keys so Python can assert they are accessible.
        auto val_ascii = kvs->get_value_f64("utf8_ascii_key");
        if (!val_ascii.has_value()) {
            throw std::runtime_error("Failed to read default value for 'utf8_ascii_key'");
        }
        auto val_greek = kvs->get_value_f64(u8"utf8_greek κλμ");
        if (!val_greek.has_value()) {
            throw std::runtime_error(u8"Failed to read default value for 'utf8_greek κλμ'");
        }
        std::cout << "default key=utf8_ascii_key value=" << val_ascii.value() << "\n";
        std::cout << u8"default key=utf8_greek κλμ value=" << val_greek.value() << "\n";

        if (!KvsInstance::normalize_snapshot_file_to_rust_envelope(params)) {
            std::cerr << "Warning: Failed to normalize snapshot file" << std::endl;
        }
    }
};

/// Read a default value whose key is a UTF-8 emoji string without ever calling
/// set_value on it. Write the retrieved value to an ASCII result key and flush.
/// Python verifies the result key equals the expected default, combining
/// feat_req__persistency__default_value_get with
/// feat_req__persistency__support_datatype_keys in one observable storage outcome.
class Utf8DefaultValueGet : public Scenario {
public:
    std::string name() const override {
        return "utf8_default_value_get";
    }

    void run(const std::string& input) const override {
        KvsParameters params = KvsParameters::from_json_section(input, "kvs_parameters_1");

        auto kvs_opt = KvsInstance::create(params);
        if (!kvs_opt) {
            throw std::runtime_error("Failed to create KVS instance");
        }
        auto kvs = *kvs_opt;

        // Read the default via a UTF-8 emoji key — never explicitly set.
        auto default_result = kvs->get_value_f64(u8"probe 🔍");
        if (!default_result.has_value()) {
            throw std::runtime_error(u8"Failed to read default value for 'probe 🔍'");
        }

        // Persist to an ASCII result key so Python can verify without UTF-8 key lookup.
        if (!kvs->set_value("result_key", default_result.value())) {
            throw std::runtime_error("Failed to set result_key");
        }
        if (!kvs->flush()) {
            throw std::runtime_error("Failed to flush KVS");
        }
        if (!KvsInstance::normalize_snapshot_file_to_rust_envelope(params)) {
            std::cerr << "Warning: Failed to normalize snapshot file" << std::endl;
        }
    }
};

}  // anonymous namespace

Scenario::Ptr make_utf8_defaults_scenario() {
    return std::make_shared<Utf8Defaults>();
}

Scenario::Ptr make_utf8_default_value_get_scenario() {
    return std::make_shared<Utf8DefaultValueGet>();
}

File: feature_integration_tests/test_scenarios/rust/src/scenarios/persistency/utf8_defaults.rs
// *******************************************************************************
// Copyright (c) 2026 Contributors to the Eclipse Foundation
//
// See the NOTICE file(s) distributed with this work for additional
// information regarding copyright ownership.
//
// This program and the accompanying materials are made available under the
// terms of the Apache License Version 2.0 which is available at
// <https://www.apache.org/licenses/LICENSE-2.0>
//
// SPDX-License-Identifier: Apache-2.0
// *******************************************************************************
use crate::internals::persistency::{kvs_instance::kvs_instance, kvs_parameters::KvsParameters};
use rust_kvs::prelude::KvsApi;
use serde_json::Value;
use test_scenarios_rust::scenario::Scenario;
use tracing::info;

fn parse_params(input: &str) -> Result<KvsParameters, String> {
    let v: Value = serde_json::from_str(input).map_err(|e| e.to_string())?;
    KvsParameters::from_value(&v["kvs_parameters_1"]).map_err(|e| e.to_string())
}

/// Open a KVS whose defaults file was created with UTF-8 encoded key names.
/// Override only the emoji key and flush. Python verifies:
/// - the emoji key appears in the snapshot with the override value
/// - the ASCII and Greek keys are absent from the snapshot (never explicitly written)
///
/// Combines feat_req__persistency__support_datatype_keys,
/// feat_req__persistency__default_values, and
/// feat_req__persistency__default_value_file.
pub struct Utf8Defaults;

impl Scenario for Utf8Defaults {
    fn name(&self) -> &str {
        "utf8_defaults"
    }

    fn run(&self, input: &str) -> Result<(), String> {
        let params = parse_params(input)?;
        let kvs = kvs_instance(params).map_err(|e| format!("{e:?}"))?;

        // Override only the emoji key; ASCII and Greek keys are intentionally left as defaults.
        kvs.set_value("utf8_emoji 🔑", 777.0_f64)
            .map_err(|e| format!("{e:?}"))?;
        kvs.flush().map_err(|e| format!("{e:?}"))?;
        // Log default values for ASCII and Greek keys so Python can assert accessibility.
        let val_ascii: f64 = kvs
            .get_value_as("utf8_ascii_key")
            .map_err(|e| format!("Failed to read default utf8_ascii_key: {e:?}"))?;
        let val_greek: f64 = kvs
            .get_value_as("utf8_greek κλμ")
            .map_err(|e| format!("Failed to read default utf8_greek κλμ: {e:?}"))?;
        info!(key = "utf8_ascii_key", value = val_ascii, source = "default");
        info!(key = "utf8_greek κλμ", value = val_greek, source = "default");
        Ok(())
    }
}

/// Read a default value whose key is a UTF-8 emoji string without ever calling
/// set_value on it. Write the retrieved value to an ASCII result key and flush.
/// Python verifies the result key equals the expected default, combining
/// feat_req__persistency__default_value_get with
/// feat_req__persistency__support_datatype_keys in one observable storage outcome.
pub struct Utf8DefaultValueGet;

impl Scenario for Utf8DefaultValueGet {
    fn name(&self) -> &str {
        "utf8_default_value_get"
    }

    fn run(&self, input: &str) -> Result<(), String> {
        let params = parse_params(input)?;
        let kvs = kvs_instance(params).map_err(|e| format!("{e:?}"))?;
        // Read the default value via a UTF-8 emoji key — never explicitly set.
        let default_val: f64 = kvs
            .get_value_as("probe 🔍")
            .map_err(|e| format!("Failed to read UTF-8 default value: {e:?}"))?;
        // Persist to an ASCII result key so Python can verify without UTF-8 key lookup.
        kvs.set_value("result_key", default_val).map_err(|e| format!("{e:?}"))?;
        kvs.flush().map_err(|e| format!("{e:?}"))?;
        Ok(())
    }
}
